### PR TITLE
Add support for scheme pre-actions and post-actions

### PR DESCRIPTION
**Pull request by kastiglione**
## Add support for scheme pre-actions and post-actions

This pull request adds `preActions` and `postActions` to each of the [`scheme`](https://github.com/yonaskolb/XcodeGen/blob/178723101711d722151a0d16fcbff75233785b63/Docs/ProjectSpec.md#scheme) action types. For example:

```yaml
schemes:
  Foo:
    build:
      preActions:
        - name: Prepare Build
          script: echo Preparing for Build
```

✅ **Check List** ✅

- [x] Add tests
- [x] Add some actions to the scheme spec loading test
- [x] Add a custom scheme with actions to the test fixture
- [x] Add support for inheriting build settings (see [`environmentBuildable`](https://github.com/xcodeswift/xcproj/blob/a2812ce567434fc4f9ba287ad341756e006be274/Sources/xcproj/XCScheme.swift#L175))
- [x] Update ProjectSpec.md
- [x] Check whether `script` needs to be escaped

This change requires xcodeswift/xcproj#217.

### CODE REVIEW

**Comment by kastiglione** [comment]
@yonaskolb To build XcodeGen with the latest HEAD of xcproj, I had to make the changes in 1fe2e96. Feel free to cherrypick it into #227.

**Comment by yonaskolb** [comment]
@kastiglione, #227 is merged

**Comment by yonaskolb** [comment]
Do you want to add pre and post actions to `TargetScheme` as well? It would just apply them to all the different actions in the Scheme init here https://github.com/yonaskolb/XcodeGen/blob/master/Sources/XcodeGenKit/ProjectGenerator.swift#L178

**Comment by kastiglione** [comment]
I don't need that functionality for my work. Do you think it will be common enough for people to have a script they want executed before/after every build, run, test, profile, and archive? I don't know the answer to this, but I can check, are these scripts called with arguments or environment variables that indicate which action (build/run/test/etc) is being invoked? If not, then I would say not to add it.

**Comment by kastiglione** [comment]
I just checked, the pre/post action scripts don't have any arguments or environment variables that tell which type of action (build/run/test/etc) triggered the script.

**Comment by kastiglione** [comment]
Thanks for the review and approval. I plan to do the changes mentioned in the todo checklist. That can be here or a follow up.

**Comment by yonaskolb** [comment]
They can be done here 👍

**Comment by kastiglione** [comment]
Ready to merge, pending review of updates.

**Comment by yonaskolb** [comment]
Looks great @kastiglione, thanks!
If you have time to add some actions to the scheme spec loading test, and add a custom scheme with actions to the test fixture that would be great

**Comment by kastiglione** [comment]
Tests added.

**Comment by yonaskolb** [comment]
Nice! Just the updated test project needs to be committed after running the tests. The tests won't fail automatically with scheme changes at the moment due to https://github.com/yonaskolb/XcodeGen/issues/202

File: Tests/XcodeGenKitTests/FixtureTests.swift [hunk 1]
```diff
@@ -59,5 +59,15 @@ func fixtureTests() {
                 try expect(variableGroup.children.filter { $0 == refs.first?.reference }.count) == 1
             }
         }
+
+        $0.it("generates scheme execution actions") {
```
Comment (yonaskolb) [inline Tests/XcodeGenKitTests/FixtureTests.swift]: As a note for the future, tests here aren't really required as if the generated project changes the test will fail. Someone just added some more explicit checks afterwards above. No harm in having this here though 👍

**Comment by yonaskolb** [comment]
I'll do it in master 👍

**Pull request merged**